Split into separate files and add unit tests

jcansdale opened the pull request:
A while back I was doing some work on VsixUtil, but finding it tricky to change stuff without worrying about breaking it. I ended up splitting the source into separate files and adding a bunch of unit tests. I realize this is a big change, so feel free to tell me where to stick my PR. ;-)

If you were to accept it, it would certainly make it easier to submit PRs in future.

Do you still use this utility yourself at all?

Cheers,
Jamie.

jcansdale committed:
Merge pull request #1 from jcansdale/extended-sku

Install for all specified SKUs

jcansdale committed:
Merge pull request #2 from jcansdale/extended-sku

Bump solution version

jcansdale committed:
Separate source files

jcansdale committed:
Add tests for `Program.ParseCommandLine`

jcansdale committed:
Move assembly resolving into ApplicationContext

jcansdale committed:
Marshal console output into default app domain

jcansdale committed:
Factor out ExtensionManagerFactory

jcansdale committed:
Check it works when `Microsoft.VisualStudio.ExtensionManager` is local

Having `Microsoft.VisualStudio.ExtensionManager` in output directory is similar to being in GAC (which we want to test).

jcansdale committed:
Find VS 2017 instances using SetupConfiguration

Remove SKU option. It was a hack that only worked with VS 2017 instances installed in the default location.

jcansdale committed:
Use full Version and VsVersion enum

jcansdale committed:
Run tests against all installed versions

jcansdale committed:
Use VsVersion everywhere

jcansdale committed:
Add product filter and tests

All extensions to be installed for specific product (e.g. Community or Enterprise).

jcansdale committed:
Use `IConsoleContext` for all console output

Use `CommandRunnerFactory` and `CommandRunner` as remote objects.

jcansdale committed:
Use InstalledVersion instead of AppPath / VsVersion

jcansdale committed:
Add support for years as version number

jcansdale commented on Src/Program.cs:
```diff
@@ -1,490 +1,20 @@
-            var version = "";
-            var sku = "Community;Professional;Enterprise";
+            VsVersion? version = null;
+            string product = null;
```
It will now install on all SKUs by default.

grokys commented on Src/VsVersion.cs:
```diff
@@ -0,0 +1,11 @@
+﻿namespace VsixUtil
+{
+    public enum VsVersion
```
Hey Jamie, apologies for the noise. Testing PR comments.

jcansdale commented on Src/Program.cs:
```diff
@@ -497,7 +27,7 @@ private static CommandLine ParseCommandLine(string[] args)
                         if (index + 1 >= args.Length)
                         {
-                            Console.Write("/install requires an argument");
+                            consoleContext.Write("/install requires an argument");
```
This makes it easy to test console output.

jcansdale commented on Src/Program.cs:
```diff
@@ -566,84 +96,69 @@ private static CommandLine ParseCommandLine(string[] args)
+            consoleContext.WriteLine("vsixutil /list [filter]");
+            consoleContext.WriteLine("");
+            consoleContext.WriteLine("The following filters can be used with all commands:");
+            consoleContext.WriteLine("   /version 15 | 2017");
```
Version can be specified release number or a year.

jcansdale commented on Src/Program.cs:
```diff
@@ -566,84 +96,69 @@ private static CommandLine ParseCommandLine(string[] args)
+            consoleContext.WriteLine("");
+            consoleContext.WriteLine("The following filters can be used with all commands:");
+            consoleContext.WriteLine("   /version 15 | 2017");
+            consoleContext.WriteLine("   /product com | Community");
```
Whole or partial product SKUs can be specified (e.g. com, ent, pro or team).

jcansdale commented:
Not sure why the build is broken. Does AppVeyor not automatically restore NuGet packages?

jcansdale commented on Test/VsixUtil.Tests/packages.config:
```diff
@@ -0,0 +1,5 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<packages>
+  <package id="Microsoft.VisualStudio.Setup.Configuration.Interop" version="1.8.24" targetFramework="net452" developmentDependency="true" />
```
I'm not sure why this isn't being restored in CI?

jaredpar commented:
@jcansdale 

> Not sure why the build is broken. Does AppVeyor not automatically restore NuGet packages?

Looks like that's actually the problem. The appveyor.xml file I have in this repo doesn't have a restore step. You'll need to add the following section:

``` yaml
before_build: 
  # Updates the version number in the .vsixmanifest and updates the AppVeyor build number to match
  - nuget restore VsixUtil.sln
```

jcansdale committed:
Add `nuget restore` to before_build

jcansdale commented:
Still trying to work out why the build is failing. 😕 

I notice there is a branch called `pr/6-split-into-separate-files-and-add-unit-tests`. Was that you or is this a GitHub thing?

jcansdale committed:
Revert back to VS 2015 solution

I've been using VS2017, but AppVeyor compiles with MSBuild 14.0.

jcansdale committed:
Fix when building with MSBuild 14.0

Not sure why it worked in VS 2015 but not MSBuild 14.0.

jcansdale commented:
Okay, so I need to fix the tests now. Yay for unit tests. 😄

jcansdale committed:
Don't fail when VS 2017 isn't installed

`new SetupConfiguration()` was throwing `COMException` with `REGDB_E_CLASSNOTREG`.
`GetSetupInstances` now returns no instances when VS 2017 isn't installed.
